tfe/data_source_outputs.go

```diff
@@ -0,0 +1,234 @@
+			},
+		}, nil
+	}
+	return &tfprotov5.ReadDataSourceResponse{
+		State: &state,
+	}, nil
+}
+
+func (d dataSourceOutputs) ValidateDataSourceConfig(ctx context.Context, req *tfprotov5.ValidateDataSourceConfigRequest) (*tfprotov5.ValidateDataSourceConfigResponse, error) {
+	return &tfprotov5.ValidateDataSourceConfigResponse{}, nil
+}
+
+func (d dataSourceOutputs) readConfigValues(req *tfprotov5.ReadDataSourceRequest) (string, string, error) {
+	var orgName string
+	var wsName string
+	var err error
+
+	config := req.Config
+	val, err := config.Unmarshal(tftypes.Object{
+		AttributeTypes: map[string]tftypes.Type{
+			"workspace":    tftypes.String,
+			"organization": tftypes.String,
+			"values":       tftypes.DynamicPseudoType,
+			"id":           tftypes.String,
+		}})
+	if err != nil {
+		return orgName, wsName, fmt.Errorf("Error unmarshalling config: %v", err)
+	}
+
+	var valMap map[string]tftypes.Value
+	err = val.As(&valMap)
+	if err != nil {
+		return orgName, wsName, fmt.Errorf("Error assigning configuration attributes to map: %v", err)
+	}
+
+	if valMap["organization"].IsNull() || valMap["workspace"].IsNull() {
+		return orgName, wsName, fmt.Errorf("Organization and Workspace cannot be nil: %v", err)
+	}
+
+	err = valMap["organization"].As(&orgName)
+	if err != nil {
+		return orgName, wsName, fmt.Errorf("Error assigning 'organization' value to string: %v", err)
+	}
+
+	err = valMap["workspace"].As(&wsName)
+	if err != nil {
+		return orgName, wsName, fmt.Errorf("Error assigning 'workspace' value to string: %v", err)
+	}
+
+	return orgName, wsName, nil
+}
+
+type stateData struct {
+	outputs map[string]*outputData
+}
+
+type outputData struct {
+	Value cty.Value
+}
+
+func (d dataSourceOutputs) readStateOutput(ctx context.Context, tfeClient *tfe.Client, orgName, wsName string) (*stateData, error) {
+	log.Printf("[DEBUG] Reading the Workspace %s in Organization %s", wsName, orgName)
+	ws, err := tfeClient.Workspaces.Read(ctx, orgName, wsName)
+	if err != nil {
+		return nil, fmt.Errorf("Error reading workspace: %v", err)
+	}
+
+	log.Printf("[DEBUG] Reading the current StateVersion for Workspace ID %s.", ws.ID)
+	sv, err := tfeClient.StateVersions.Current(ctx, ws.ID)
+	if err != nil {
+		if err == tfe.ErrResourceNotFound {
+			return nil, fmt.Errorf("Current state version for workspace '%s' not found.", wsName)
+		}
+		return nil, fmt.Errorf("Could not read the current state for workspace '%s' : %v", wsName, err)
+	}
+
+	opts := tfe.StateVersionOutputsListOptions{}
+	outputs, err := tfeClient.StateVersions.Outputs(ctx, sv.ID, opts)
+	if err != nil {
+		return nil, fmt.Errorf("Could not read the outputs for state version '%s': %v", sv.ID, err)
+	}
+
+	sd := &stateData{
+		outputs: map[string]*outputData{},
+	}
+
+	for _, op := range outputs {
+		buf, err := json.Marshal(op.Value)
+		if err != nil {
+			return nil, fmt.Errorf("Could not marshal output value: %v", err)
+		}
+
+		v := ctyjson.SimpleJSONValue{}
+		err = v.UnmarshalJSON(buf)
```

Comment:
Wow really? I'm sort of surprised it's as easy as using JSON marshalling; I've never really considered how Terraform types and JSON types appear to be truly 1:1.